File: Cake List/NetworkService.m
//
//  NetworkService.m
//  Cake List
//
//  Created by Harmeet Singh on 05/04/2018.
//  Copyright © 2018 Stewart Hart. All rights reserved.
//

#import "NetworkService.h"
#import "NSError+Extension.h"
#import "Cake.h"

@implementation NetworkService

- (void)fetchAllCakesWithCompletion:(FetchCakesCompletion)completion {
    
    NSURL *url = [NSURL URLWithString:@"https://gist.githubusercontent.com/hart88/198f29ec5114a3ec3460/raw/8dd19a88f9b8d24c23d9960f3300d0c917a4f07c/cake.json"];
    
    NSURLSession *session = [NSURLSession sharedSession];
    __weak typeof(self) weakSelf = self;
    
    [[session dataTaskWithURL:url completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        
        if (error) {
            
            return [weakSelf requestCompletedWithCakes:nil error:error completion:completion];
            
        } else if (data && response) {
            
            return [weakSelf cakesRequestCompletedWithResponse:response data:data completion:completion];
        }
        
        NSError *unkownError = [NSError errorWithMessage:@"An unkown error occured. Please try again later."];
        [weakSelf requestCompletedWithCakes:nil error:unkownError completion:completion];
        
    }] resume] ;
}

- (void)cakesRequestCompletedWithResponse:(NSURLResponse *)response data:(NSData *)data completion:(FetchCakesCompletion)completion {
    
    NSHTTPURLResponse *httpURLResponse = (NSHTTPURLResponse *)response;
    BOOL isSuccessfulRequest = httpURLResponse.statusCode >= 200 && httpURLResponse.statusCode < 400;
    
    if (!isSuccessfulRequest) {
        
        NSError *requestFailedError = [NSError errorWithMessage:@"The network request failed. Please try again later"];
        return [self requestCompletedWithCakes:nil error:requestFailedError completion:completion];
    }
    
    NSError *jsonError;
    id responseData = [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:&jsonError];
    
    if (jsonError) {
        
        return [self requestCompletedWithCakes:nil error:jsonError completion:completion];
    
    } else if ([responseData isKindOfClass:[NSArray class]]) {
        
        NSArray *cakesArray = (NSArray *)responseData;
        NSMutableArray *allCakes = [NSMutableArray new];
        
        for (NSDictionary *cakeDictionary in cakesArray) {
            
            Cake *cake = [[Cake alloc] initWithDictionary:cakeDictionary];
            [allCakes addObject:cake];
        }
        
        return [self requestCompletedWithCakes:allCakes error:nil completion:completion];
    }

    NSError *unkownResponseDataError = [NSError errorWithMessage:@"The server returned unexpected data. Please try again later."];
    [self requestCompletedWithCakes:nil error:unkownResponseDataError completion:completion];
}

- (void)requestCompletedWithCakes:(NSArray <Cake *>*)cakes error:(NSError *)error completion:(FetchCakesCompletion)completion {
    
    dispatch_async(dispatch_get_main_queue(), ^{
        
        completion(cakes, error);
    });
}

@end

File: Cake List/NetworkService.swift
//
//  NetworkService.swift
//  Cake List
//
//  Created by Harmeet Singh on 05/04/2018.
//  Copyright © 2018 Stewart Hart. All rights reserved.
//

import Foundation
import UIKit

class NetworkService: NSObject {
    
    // MARK: Properties
    
    typealias FetchRequestCompletion<T> = (T?, Error?) -> Swift.Void
    typealias FetchCakesCompletion = FetchRequestCompletion<[Cake]>
    typealias FetchCakeImageCompletion = FetchRequestCompletion<UIImage>
    
    
    fileprivate let session = URLSession.shared
    
    // MARK: Instantiation
    
    static let shared = NetworkService()
    
    // MARK: Requests
    
    func fetchAllCakes(with completion: @escaping FetchCakesCompletion) {
        
        guard let cakesURL = Constants.CakesURL else {
            
            return completion(nil, nil)
        }
        
        fetchData(for: cakesURL, completion: completion)
    }
    
    func fetchImage(for cake: Cake, completion: @escaping FetchCakeImageCompletion) {
        
        guard let imageURLString = cake.imageURLString, let imageURL = URL(string: imageURLString) else {
            
            return completion(nil, nil)
        }
        
        fetchData(for: imageURL, completion: completion)
    }
}

// MARK: Request implementation

extension NetworkService {
    
    func fetchData<T>(for url: URL, completion: @escaping FetchRequestCompletion<T>) {
        
        session.dataTask(with: url) { (data: Data?, response: URLResponse?, error: Error?) in
            
            if let error = error {
                
                return self.requestCompleted(with: nil, error: error, completion: completion)
            }
            
            if let response = response {
                
                if response.isSuccessfulRequest() {
                    
                    return self.parseResponse(for: data, completion: completion)
                }
                
                return self.requestCompleted(with: nil, error: NetworkServiceError.networkRequestFailed, completion: completion)
            }
            
            self.requestCompleted(with: nil, error: NetworkServiceError.unknownError, completion: completion)
      
        }.resume()
    }
    
    func parseResponse<T>(for data: Data?, completion: @escaping FetchRequestCompletion<T>) {
        
        guard let data = data else {
            
            return requestCompleted(with: nil, error: NetworkServiceError.responseDataNil, completion: completion)
        }
        
        if completion is FetchCakesCompletion {
            
            do {
                
                var allCakes: [Cake] = []
                let jsonObject = try JSONSerialization.jsonObject(with: data, options: .allowFragments) as? [[String:AnyObject]]
                
                for cakeDictionary in jsonObject! {
                    
                    let cake = Cake(dictionary: cakeDictionary)
                    allCakes.append(cake!)
                }
                
                return requestCompleted(with: allCakes as? T, error: nil, completion: completion)
                
            } catch let serializationError {
                
                return requestCompleted(with: nil, error: serializationError, completion: completion)
            }
        
        } else if completion is FetchCakeImageCompletion {
            
            let cakeImage = UIImage(data: data)
            return requestCompleted(with: cakeImage as? T, error: nil, completion: completion)
        }
    }
    
    func requestCompleted<T>(with object: T?, error: Error?, completion: @escaping FetchRequestCompletion<T>) {
        
        DispatchQueue.main.async {
            
            completion(object, error)
        }
    }
}
